posts/lessons-learned-ai-automation-rhel/index.md

```diff
@@ -0,0 +1,160 @@
+    GitLogSearchTool(options=local_tool_options),
+    GitPreparePackageSourcesTool(options=local_tool_options),
+    # ... more specialized tools
+]
+```
+
+**Lesson learned:** Invest time in tooling and architecture.
+
+## Mixing AI creativity with code reliability
+
+This might be the most important lesson: AI agents will never be 100% deterministic, and that's actually okay for many use cases. But you need to be strategic about where you accept non-determinism and where you demand consistency.
+
+**Examples where we embrace AI non-determinism:**
+
+- Analysing natural language in Jira issues
+- Writing commit messages and documentation
+- Searching for fixes upstream
+
+**Examples where we use deterministic code:**
+
+- Git operations (clone, checkout, apply patches)
+- File system operations
+- Complex deterministic logic, like mapping RHEL versions to branches
+
+**Lesson learned:** Design your system to be "AI-assisted" rather than "AI-controlled." Use AI for the creative, analytical work, but rely on deterministic code for the critical operations.
+
+## From code to service
+
+Once you move from building a prototype to building a production system, you also need to think about the infrastructure and other components, not just agents and code around them. Here are just a few examples of what we plumbed that keeps AI agents running reliably:
+
+- **Queue-based architecture:** We use valkey queues to orchestrate work between agents, a reliable pattern we already knew worked. Issues flow through different queues with automatic retry logic.
+- **Observability:** Every agent action flows through [Phoenix tracing](https://phoenix.arize.ai/), giving us complete visibility into what agents are thinking and doing. When something goes wrong, we can trace exactly which decisions led to the failure. This isn't just logging - it's understanding why your AI system made specific choices.
+- **Container deployment:** The entire system runs on OpenShift and each agent type has its own deployment with resource limits. This also makes the whole system easily scalable.
+- **Dry-run capabilities:** Every operation has dry-run support built in. This helps us to better test and debug the agent's behaviour. When an agent makes an unexpected decision, we can replay the scenario safely.
+
+**Lesson learned:** Start thinking about operational capabilities on day one. Dry-run modes, proper logging, and failure recovery aren't "nice to haves" - they're what make AI systems actually usable. And since LLMs produce a ton of output, you'll need an efficient way to find out why the agents made such an odd decision.
+
+## Takeaways
+
+After roughly 3.5 months of development, our system is currently running for a selected set of components - and it's being heavily developed. But by following these principles, we've seen a huge difference in reliability and maintainability. More importantly, we're now much clearer about what's actually possible to do with AI and what isn't.
```

Comment:
> More importantly, we're now much clearer about what's actually possible to do with AI and what isn't.

I know what you're trying to say, but it doesn't feel right and at the same time I don't have a better suggestion for this :/ 

---
I'll think about it a bit…